keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,6 +1069,161 @@ the remaining requests in that queue start getting faster service.  In
+with something relatively simple to first get operational experience with it
+before investing into sophisticated algorithms or exposing a knob to users.
+
+In order to determine the function that will be approximating the `width` of
+a request, we should first estimate how expensive a particular request is.
+And we need to think about both dimensions that we're trying to protect from
+overloading (CPU and RAM) and how many concurrency units a request can actually
+consume.
+
+Let's start with CPU.  The total cost of processing a LIST request should be
+proportional to the number of processed objects.  However, given that in
+practice processing a single request isn't parallelized (and the fact that
+we generally scale the number of total concurrency units linearly with amount
+of available resources), a single request should consume no more than A
+concurrency units.  Fortunately that all compiles together because the
+`processing latency` of the LIST request is actually proportional to the
+number of processed objects, so the cost of the request (defined above as
+`<width> x <processing latency>` really is proportaional to the number of
+processed objects as expected.
+
+For RAM the situation is actually different.  In order to process a LIST
+request we actually store all objects that we process in memory. Given that
+memory is uncompressable resource, we effectively need to reserve all that
+memory for the whole time of processing that request.  That suggests that
+the `width` for the request from the RAM perspective should be proportional
+to the number of processed items.
+
+So what we get is that:
+```
+  width_cpu(N) = A
+  width_ram(N) = B * N
+```
+where N is the number of items a given LIST request is processing.
+
+The question is how to combine them to a single number.  To stay on the safe
+side, we can just protect every resource independently, which translates to
+to the following function:
+```
+  width(n) = max(A, B * N)
+```
+In practise, I expect it to be generally dominated by A, but the some
+approximations for those values will be figured out during tuning phase.
+
+Given that the estimation is well separated piece of logic, we can decide
+to replace with much more sophisticated logic later (e.g. whether it is
+served from etcd or from cache, whether it is namespaced or not, etc.).
+
+One more important aspect to resolve is what happens if a given priority
+level doesn't have enough concurrency units assigned to it. To be on the
+safe side we should probably implement borrowing across priority levels.
+However, given we don't want to block introducing the `width` concept on
+design and implementation of borrowing, until this is done we have two
+main options:
+- cap the `width` at the concurrency units assigned to the priority level
+- reject requests for which we won't be able to allocate enough concurrency
+  units
+To avoid breaking some users, we will proceed with the first option.
```

Comment:
I agree with this, but also somewhere I wrote up an idea around how to pay for excess width by forcing extra duration between requests, so that it devolves into rate limiting / a duty cycle.

Reply:
We were discussing this for watch. And it's already mentioned below for the watch (see line 1400 below).

I agree the same approach here makes a lot of sense - I will add that tomorrow.

Reply:
I updated the doc to reflect this - PTAL